### DISCUSSION TITLE
AdmissionWebhooks for non compliant RTMP streams

### DISCUSSION THREAD

**Discussion by wizeyyy** [OvenMediaLabs/OvenMediaEngine#1170] [Help] [answered]
I have some non compliant RTMP streams in the form of (rtmp://host[:port]/app/stream/channel/file) and other complaint ones in the form of (rtmp://host[:port]/app/stream/file)

I am trying to use AdmissionWebhooks to change the url, however when a noncompliant stream starts I get the following error in the log

[2023-04-04 14:00:37.200] I [SPRTMP-t1935:22898] RTMPProvider | rtmp_provider.cpp:189  | A RTMP client has connected from <ClientSocket: 0x7f5f2d401010, #5, Connected, TCP, Nonblocking, 41.13.18.165:49231>
[2023-04-04 14:00:37.525] E [SPRTMP-t1935:22898] RTMPProvider | rtmp_stream.cpp:276  | #access#stream#1 application does not exist
[2023-04-04 14:00:37.525] I [SPRTMP-t1935:22898] Provider | stream.cpp:63   | Unknown/(101) has been stopped playing stream
[2023-04-04 14:00:37.525] I [SPRTMP-t1935:22898] RTMPProvider | rtmp_provider.cpp:218  | The RTMP client has disconnected: [Unknown/], remote: <ClientSocket: 0x7f5f2d401010, #5, Closed, TCP, Nonblocking, 41.13.18.165:49231>

My config looks as follows

```
<VirtualHost>
                        <Name>access</Name>
   
                        <Distribution>ovenmediaengine.com</Distribution>

                           <Host>
                                <Names>

                                      
						<Name>admit.*****</Name>
                                </Names>
                           </Host>

                        <AdmissionWebhooks>
                                <ControlServerUrl>http://play.******:8080/access.php</ControlServerUrl>
                                <SecretKey>1234</SecretKey>
                                        <Timeout>3000</Timeout>
                                        <Enables>
                                                <Providers>rtmp</Providers>
                                                
                                        </Enables>
                        </AdmissionWebhooks>
</VirtualHost>
```

I can get the complaint ones to work with the AdmissionWebhooks if I add an application into the virtual host for the AdmissionWebhooks

The application has an RTMP Provider and RTMPPush Publisher

Then the webhook can redirect it to completely different virtualhost on OME.
But I was hopping to achieve, that any RTMP url can come through and the Webhook would change the url to another Virtualhost on OME.

The ControlServerUrl returns the following

```
{
  		"allowed": true,
  		"new_url": "rtmp://play2.*****:1935/1/862******",
  		"lifetime": 0,
  		"reason": "authorized"}
```

**Comment by wizeyyy** [comment]
This question is quite important because there are many Chinese devices that are hitting the market and are using rtmp streams in the form of rtmp://host[:port]/app/stream/channel/file, i.e they re adding an extra parameter called channel.

AdmissionWebHooks would work fine if it didn't need to match the orginal stream with an application, and rather redirect it to the new url.
The documentation seems to indicate that this is how it should work. But in practice this isn't happening.

**Comment by getroot** [comment, marked as answer]
Because RTMP was validating the application before AdmissionWebhooks, it didn't work as intended. I have corrected it. (28085d80fd09d56d50c2c13fc92c96e43a7feee6)

Also, now sending an RTMP stream to the `rtmp://host:port/app/channel/stream` endpoint will create a stream to `host#app/channel` .

**Comment by wizeyyy** [comment]
Thanks will test shortly

**Comment by wizeyyy** [comment]
Thanks it works well
